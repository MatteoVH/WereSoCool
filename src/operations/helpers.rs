//#[allow(dead_code)]
//pub mod helpers {
//    use event::Event;
//
//    pub fn vv_event_to_v_events(vv_events: &Vec<Vec<Event>>) -> Vec<Event> {
//        let mut accumulator = vec![];
//        let mut state = vv_events.clone();
//        while state.len() > 0 {
//            state.retain(|ref x| x.len() > 0);
//            fold_vv_events(&mut accumulator, &mut state);
//        }
//        accumulator
//    }
//
//    fn fold_vv_events(accumulator: &mut Vec<Event>, state: &mut Vec<Vec<Event>>) {
//        let next_length = next_length(&state);
//        let mut s = state.clone();
//        let mut events_to_join = vec![];
//        for (index, vec_event) in state.iter_mut().enumerate() {
//            if vec_event.len() > 0 {
//                let current = vec_event.remove(0);
//                if current.length > next_length {
//                    let mut remainder = current.clone();
//                    remainder.length = current.length - next_length;
//                    vec_event.insert(0, remainder)
//                }
//                events_to_join.push(current);
//            } else {
//                s.remove(index);
//            }
//        }
//
//        if events_to_join.len() > 0 {
//            let event = join_events(events_to_join, next_length);
//            if event.length > 0.005 {
//                accumulator.push(event)
//            }
//        }
//    }
//
//    fn join_events(events: Vec<Event>, length: f32) -> Event {
//        let mut sounds = vec![];
//
//        for mut event in events {
//            sounds.append(&mut event.sounds)
//        }
//
//        Event { sounds, length }
//    }
//
//    fn next_length(state: &Vec<Vec<Event>>) -> f32 {
//        let mut values = vec![];
//        for vec_event in state.iter() {
//            values.push(vec_event[0].length)
//        }
//        let min = values.iter().cloned().fold(1.0 / 0.0, f32::min);
//
//        if min.is_infinite() {
//            0.0
//        } else {
//            min
//        }
//    }
//}
