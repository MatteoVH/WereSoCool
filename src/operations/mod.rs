use event::Event;
mod apply;
mod apply_with_order;
mod get_length_ratio;
mod helpers;

#[derive(Clone, PartialEq, Debug)]
pub enum Op {
    AsIs,
    Reverse,
    PanM {
        m: f32,
    },
    PanA {
        a: f32,
    },
    TransposeM {
        m: f32,
    },
    TransposeA {
        a: f32,
    },
    Silence {
        m: f32,
    },
    Length {
        m: f32,
    },
    Gain {
        m: f32,
    },
    //    Capture { n: usize },
    Repeat {
        n: usize,
        operations: Vec<Op>,
    },
    Sequence {
        operations: Vec<Op>,
    },
    Compose {
        operations: Vec<Op>,
    },
    ComposeWithOrder {
        operations: Vec<Op>,
        order_fn: fn(usize) -> f32,
    },
    Fit {
        with_length_of: Box<Op>,
        main: Box<Op>,
        n: usize,
    },
    Overlay {
        operations: Vec<Op>,
    },
}

pub trait Apply {
    fn apply(&self, events: Vec<Event>) -> Vec<Event>;
}

pub trait ApplyWithOrder {
    fn apply_with_order(&self, order_fn: fn(order: usize) -> f32, events: Vec<Event>)
        -> Vec<Event>;
}

pub trait GetLengthRatio {
    fn get_length_ratio(&self) -> f32;
}

#[cfg(test)]
mod test;
