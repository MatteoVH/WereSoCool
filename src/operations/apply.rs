pub mod apply {
    extern crate num_rational;
    use event::Event;
    use generation::parsed_to_waveform::r_to_f32;
    use operations::helpers::helpers::vv_event_to_v_events;
    use operations::{Apply, GetLengthRatio};
    use socool_parser::ast::Op;

    impl Apply for Op {
        fn apply(&self, events: Vec<Event>) -> Vec<Event> {
            let mut vec_events: Vec<Event> = vec![];
            match self {
                Op::AsIs {} => {
                    vec_events = events;
                }

                Op::Reverse {} => {
                    let mut clone = events.clone();
                    clone.reverse();
                    vec_events = clone
                }

                Op::TransposeM { m } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.frequency = sound.frequency * r_to_f32(*m);
                        }
                        vec_events.push(e)
                    }
                }

                Op::TransposeA { a } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.frequency = sound.frequency + r_to_f32(*a);
                        }
                        vec_events.push(e)
                    }
                }

                Op::PanA { a } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.pan += r_to_f32(*a);
                        }
                        vec_events.push(e)
                    }
                }

                Op::PanM { m } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.pan *= r_to_f32(*m);;
                        }
                        vec_events.push(e)
                    }
                }

                Op::Length { m } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        e.length = e.length * r_to_f32(*m);;
                        vec_events.push(e)
                    }
                }

                Op::Silence { m } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        e.length *= r_to_f32(*m);
                        for sound in e.sounds.iter_mut() {
                            sound.frequency = 0.0;
                            sound.gain = 0.0;
                        }
                        vec_events.push(e)
                    }
                }

                Op::Gain { m } => {
                    for event in events.iter() {
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.gain *= r_to_f32(*m)
                        }
                        vec_events.push(e)
                    }
                }

                Op::Compose { operations } => {
                    let mut es = events.clone();
                    for operation in operations.iter() {
                        es = operation.apply(es);
                    }
                    vec_events = es;
                }

                Op::Sequence { operations } => {
                    let mut es = events.clone();
                    let mut container = vec![];
                    for operation in operations.iter() {
                        container.push(operation.apply(es.clone()));
                    }

                    vec_events = container.iter().flat_map(|evt| evt.clone()).collect();
                }

                Op::Overlay { operations } => {
                    let mut vec_vec_events: Vec<Vec<Event>> = vec![];

                    for operation in operations.iter() {
                        let es = operation.apply(events.clone());
                        vec_vec_events.push(es);
                    }

                    vec_events = vv_event_to_v_events(&vec_vec_events);
                }

                Op::WithLengthRatioOf {
                    with_length_of,
                    main,
                } => {
                    let mut es = events.clone();

                    let target_length = with_length_of.get_length_ratio();
                    let main_length = main.get_length_ratio();
                    let ratio = target_length / main_length;

                    let new_op = Op::Length { m: ratio };

                    vec_events = new_op.apply(es);
                }
            }

            vec_events
        }
    }
}
