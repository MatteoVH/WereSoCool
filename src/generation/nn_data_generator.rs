use num_rational::Rational64;
use socool_ast::ast::{OpOrNf::*, OpOrNfTable, Op::*, Op};
use socool_ast::operations::{NormalForm, Normalize as NormalizeOp, PointOp};
use insta::assert_debug_snapshot_matches;

pub fn r_to_f64(r: Rational64) -> f64 {
    *r.numer() as f64 / *r.denom() as f64
}

#[derive(Debug, Clone)]
pub struct DataOp {
    fm: f64,
    fa: f64,
    pm: f64,
    pa: f64,
    g: f64,
    l: f64,
}

#[derive(Debug, Clone)]
pub struct CSVOp {
    fm: f64,
    fa: f64,
    pm: f64,
    pa: f64,
    g: f64,
    t: f64,
}

impl DataOp {
    pub fn to_csv_op(&self, time: f64) -> CSVOp {
        CSVOp {
            fm: self.fm,
            fa: self.fa,
            pm: self.pm,
            pa: self.pa,
            g: self.g,
            t: time,
        }
    }
}

pub type NormalData = Vec<Vec<DataOp>>;
pub type CSVData = Vec<Vec<CSVOp>>;

pub fn point_op_to_data_op(
    point_op: &PointOp,
) -> DataOp {
    let result = DataOp {
        fm: r_to_f64(point_op.fm),
        fa: r_to_f64(point_op.fa),
        pm: r_to_f64(point_op.pm),
        pa: r_to_f64(point_op.pa),
        g: r_to_f64(point_op.g),
        l: r_to_f64(point_op.l),
    };

    result
}

pub fn composition_to_normal_data(composition: &NormalForm, table: &OpOrNfTable) -> NormalData {
    let mut normal_form = NormalForm::init();

    println!("Generating Composition \n");
    composition.apply_to_normal_form(&mut normal_form, table);

    let result: NormalData = normal_form
        .operations
        .iter()
        .map(|vec_point_op| {
            let mut result = vec![];
            for op in vec_point_op {
                let data_op = point_op_to_data_op(op);
                result.push(data_op);
            }
            result
        })
        .collect();

    result
}

pub fn normal_data_to_csv_data(data: NormalData) -> CSVData {
    let subdivision = 0.5;
    let result: CSVData = data
        .iter()
        .map(|vec_data_op| {
            let mut remainder = 0.0;
            let mut seq_time = 0.0;
            let mut result = vec![];
            for op in vec_data_op {
                let op_time = op.l;
                let n_steps_to_push = (op_time + remainder)/subdivision;
                remainder = n_steps_to_push.fract();
                for n in 0..n_steps_to_push.floor() as usize {
                    let new_op =  op.to_csv_op(seq_time);
                    result.push(new_op);
                    seq_time += subdivision;
                }
            }
            result
        })
        .collect();

    result
}

#[test]
fn normal_form_to_normal_data_test() {
    let mut normal_form = NormalForm::init();
    let pt = OpOrNfTable::new();

    Overlay {
        operations: vec![
            Op(Sequence {
                operations: vec![
                    Op(PanA {
                         a: Rational64::new(1, 2),
                    }),
                    Op(TransposeM {
                        m: Rational64::new(2, 1),
                    }),
                    Op(Gain {
                        m: Rational64::new(1, 2),
                    }),
                    Op(Length {
                        m: Rational64::new(2, 1),
                    }),
                ],
            }),
            Op(Sequence {
                operations: vec![Op(Length {
                    m: Rational64::new(5, 1),
                })],
            }),
        ],
    }
        .apply_to_normal_form(&mut normal_form, &pt);

    let normal_data = composition_to_normal_data(&normal_form, &pt);

    assert_debug_snapshot_matches!("normal_form_to_normal_data_test", normal_data);

    let normal_data_subdivided = normal_data_to_csv_data(normal_data);
    assert_debug_snapshot_matches!("normal_form_to_subdivided_test", normal_data_subdivided);
}
