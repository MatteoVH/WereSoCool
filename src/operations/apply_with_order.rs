pub mod apply_with_order {
    use event::Event;
    use operations::helpers::helpers::vv_event_to_v_events;
    use operations::{Apply, ApplyWithOrder, GetLengthRatio, Op};

    impl ApplyWithOrder for Op {
        fn apply_with_order(
            &self,
            order_fn: fn(order: usize, length: usize) -> f32,
            events: Vec<Event>,
        ) -> Vec<Event> {
            let mut vec_events: Vec<Event> = vec![];
            match self {
                Op::AsIs {} => {
                    vec_events = events;
                }

                Op::Reverse {} => {
                    let mut clone = events.clone();
                    clone.reverse();
                    vec_events = clone
                }

                Op::TransposeM { m } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order,events.len());
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.frequency = sound.frequency * m * order_after_fn;
                        }
                        vec_events.push(e)
                    }
                }

                Op::TransposeA { a } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order, events.len());
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.frequency = sound.frequency + a * order_after_fn;
                        }
                        vec_events.push(e)
                    }
                }

                Op::PanM { m } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order, events.len());
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.pan *= m * order_after_fn;
                        }
                        vec_events.push(e)
                    }
                }

                Op::PanA { a } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order, events.len());
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.pan += a * order_after_fn;
                        }
                        vec_events.push(e)
                    }
                }

                Op::Length { m } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order, events.len());
                        let mut e = event.clone();
                        e.length = e.length * m * order_after_fn;
                        vec_events.push(e)
                    }
                }

                Op::Repeat { n, operations } => {
                    let mut repeat_container: Vec<Event> = vec![];

                    let sequence = Op::Sequence {
                        operations: operations.to_vec(),
                    }.apply(events);

                    for _ in 0..*n {
                        repeat_container.append(&mut sequence.clone());
                    }

                    vec_events = repeat_container;
                }

                Op::Silence { m } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order, events.len());
                        let mut e = event.clone();
                        e.length *= m * order_after_fn;
                        for sound in e.sounds.iter_mut() {
                            sound.frequency = 0.0;
                            sound.gain = 0.0;
                        }
                        vec_events.push(e)
                    }
                }

                Op::Gain { m } => {
                    for (order, event) in events.iter().enumerate() {
                        let order_after_fn = order_fn(order, events.len());
                        let mut e = event.clone();
                        for sound in e.sounds.iter_mut() {
                            sound.gain = sound.gain * m * order_after_fn;
                        }
                        vec_events.push(e)
                    }
                }

                Op::Compose { operations } => {
                    let mut es = events.clone();
                    for operation in operations.iter() {
                        es = operation.apply_with_order(order_fn, es);
                    }
                    vec_events = es;
                }

                Op::ComposeWithOrder {
                    order_fn,
                    operations,
                } => {
                    let mut es = events.clone();
                    for operation in operations.iter() {
                        es = operation.apply_with_order(*order_fn, es);
                    }
                    vec_events = es;
                }

                Op::Sequence { operations } => {
                    let mut es = events.clone();
                    let mut container = vec![];
                    for operation in operations.iter() {
                        container.push(operation.apply_with_order(order_fn, es.clone()));
                    }

                    vec_events = container.iter().flat_map(|evt| evt.clone()).collect();
                }

                Op::Fit {
                    with_length_of,
                    main,
                    n,
                } => {
                    let mut es = events.clone();

                    let target_length = with_length_of.get_length_ratio();
                    let main_length = main.get_length_ratio();
                    let ratio = target_length / main_length / *n as f32;

                    let mut array_of_main = vec![];
                    for _ in 0..*n {
                        array_of_main.push(*main.clone())
                    }

                    let main_sequence = Op::Sequence {
                        operations: array_of_main,
                    };

                    let new_op = Op::Compose {
                        operations: vec![main_sequence, Op::Length { m: ratio }],
                    };

                    vec_events = new_op.apply_with_order(order_fn, es);
                }

                Op::Overlay { operations } => {
                    let mut vec_vec_events: Vec<Vec<Event>> = vec![];

                    for operation in operations.iter() {
                        let es = operation.apply_with_order(order_fn, events.clone());
                        vec_vec_events.push(es);
                    }

                    vec_events = vv_event_to_v_events(&vec_vec_events);
                }
            }

            vec_events
        }
    }
}
