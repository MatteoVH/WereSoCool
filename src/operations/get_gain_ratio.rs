pub mod get_length_ratio {
    use num_rational::{Ratio, Rational64};
    use socool_parser::ast::Op;

    use operations::GetGainRatio;

    extern crate num_rational;

    impl GetGainRatio for Op {
        fn get_gain_ratio(&self) -> Rational64 {
            match self {
                Op::AsIs {}
                | Op::Sine {}
                | Op::Square {}
                | Op::Noise {}
                | Op::FInvert {}
                | Op::Reverse {}
                | Op::TransposeM { m: _ }
                | Op::TransposeA { a: _ }
                | Op::PanA { a: _ }
                | Op::PanM { m: _ }
                | Op::Length { m: _ } => Ratio::from_integer(1),

                Op::Gain { m } => *m,

                Op::Silence { m: _ } => Ratio::from_integer(0),

                Op::Sequence { operations } => {
                    let mut max = Ratio::from_integer(0);
                    for operation in operations {
                        let gr = operation.get_gain_ratio();
                        if gr > max {
                            max = gr;
                        }
                    }

                    max
                }

                Op::Compose { operations } => {
                    let mut max = Ratio::from_integer(0);
                    for operation in operations {
                        let gr = operation.get_gain_ratio();
                        if gr > max {
                            max = gr;
                        }
                    }

                    max
                }

                Op::Choice { operations } => operations[0].get_gain_ratio(),

                Op::WithLengthRatioOf {
                    with_length_of,
                    main,
                } => main.get_gain_ratio(),

                Op::WithGainRatioOf {
                    with_gain_of,
                    main: _,
                } => with_gain_of.get_gain_ratio(),

                Op::ModulateBy { operations: _ } => Ratio::from_integer(1),

                Op::Overlay { operations } => {
                    let mut max = Ratio::new(0, 1);
                    for op in operations {
                        let next = op.get_gain_ratio();
                        if next > max {
                            max = next;
                        }
                    }
                    max
                }
            }
        }
    }
}
