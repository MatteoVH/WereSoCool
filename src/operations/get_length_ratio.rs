pub mod get_length_ratio {
    use operations::{Apply, GetLengthRatio, Op};

    impl GetLengthRatio for Op {
        fn get_length_ratio(&self) -> f32 {
            match self {
                Op::AsIs {}
                | Op::Reverse {}
                | Op::TransposeM { m: _ }
                | Op::TransposeA { a: _ }
                | Op::PanA { a: _ }
                | Op::PanM { m: _ }
                | Op::Gain { m: _ } => 1.0,

                Op::Repeat { n, operations } => {
                    let mut length_ratio_of_operations = 0.0;
                    for operation in operations {
                        length_ratio_of_operations += operation.get_length_ratio();
                    }

                    length_ratio_of_operations * *n as f32
                }

                Op::Length { m } | Op::Silence { m } => *m,

                Op::Sequence { operations } => {
                    let mut new_total = 0.0;
                    for operation in operations {
                        new_total += operation.get_length_ratio();
                    }
                    new_total
                }
                Op::Compose { operations } => {
                    let mut new_total = 1.0;
                    for operation in operations {
                        new_total *= operation.get_length_ratio();
                    }
                    new_total
                }

                Op::ComposeWithOrder {
                    order_fn,
                    operations,
                } => {
                    let mut new_total = 1.0;
                    for operation in operations {
                        new_total *= operation.get_length_ratio();
                    }
                    new_total
                }

                Op::Fit {
                    with_length_of,
                    main: _,
                    n: _,
                } => with_length_of.get_length_ratio(),

                Op::Overlay { operations } => {
                    let mut max = 0.0;
                    for op in operations {
                        let next = op.get_length_ratio();
                        if next > max {
                            max = next;
                        }
                    }
                    max
                }
            }
        }
    }
}
