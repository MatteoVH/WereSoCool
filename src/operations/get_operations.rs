pub mod get_operations {
    use operations::GetOperations;
    use socool_parser::ast::Op;
    
    impl GetOperations for Op {
        fn get_operations(&self) -> Option<Vec<Op>> {
            match self {
                Op::AsIs {}
//                | Op::Reverse {}
                | Op::TransposeM { m: _ }
//                | Op::TransposeA { a: _ }
//                | Op::PanA { a: _ }
//                | Op::PanM { m: _ }
//                | Op::Gain { m: _ }
                => None,
//
//                Op::Length { m: _ } |
                Op::Silence { m: _ } => None,
//
                Op::Sequence { operations: _ }
                | Op::Compose { operations: _ }
                => None,
//
//                Op::WithLengthRatioOf {
//                    with_length_of: _,
//                    main: _,
//                } => None,

                Op::Overlay { operations } => {  Some(operations.to_vec()) }
            }
        }
    }
}
