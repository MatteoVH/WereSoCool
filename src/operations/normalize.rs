pub mod normalize {
    extern crate num_rational;
    use num_rational::{Ratio, Rational};
    use operations::{GetLengthRatio, NormalForm, Normalize};
    use socool_parser::ast::{Op, Op::*};
    use std::cmp::Ordering::{Equal, Greater, Less};

    fn fmap_point_op(new_op: Op, input: &mut NormalForm) {
        for mut voice in input {
            for mut op in voice {
                *op = Op::Compose {
                    operations: vec![op.clone(), new_op.clone()],
                };
            }
        }
    }

    impl Normalize for Op {
        fn apply_to_normal_form(&self, input: &mut NormalForm) {
            //            println!("{:?}", input);
            match self {
                Op::AsIs => {}

                Op::Reverse => {
                    for mut voice in input.iter_mut() {
                        voice.reverse();
                    }
                }

                Op::TransposeM { m } => {
                    fmap_point_op(Op::TransposeM { m: *m }, input);
                }

                Op::TransposeA { a } => {
                    fmap_point_op(Op::TransposeA { a: *a }, input);
                }

                Op::PanA { a } => {
                    fmap_point_op(Op::PanA { a: *a }, input);
                }

                Op::PanM { m } => {
                    fmap_point_op(Op::PanM { m: *m }, input);
                }

                Op::Gain { m } => {
                    fmap_point_op(Op::Gain { m: *m }, input);
                }

                Op::Length { m } => {
                    fmap_point_op(Op::Length { m: *m }, input);
                }

                Op::Silence { m } => {
                    let max_len = get_max_length_ratio(&input);

                    for voice in input.iter_mut() {
                        *voice = vec![Op::Silence { m: max_len * m }]
                    }
                }
                //
                Op::Sequence { operations } => {
                    let mut result = vec![];

                    for op in operations {
                        let mut input_clone = input.clone();
                        op.apply_to_normal_form(&mut input_clone);
                        result = join_sequence(result, input_clone);
                    }

                    *input = result
                }

                Op::Compose { operations } => {
                    for op in operations {
                        op.apply_to_normal_form(input);
                    }
                }

                Op::WithLengthRatioOf {
                    with_length_of,
                    main,
                } => {
                    let target_length = with_length_of.get_length_ratio();
                    let main_length = main.get_length_ratio();
                    let ratio = target_length / main_length;
                    println!("{:?}", ratio);
                    let new_op = Op::Length { m: ratio };

                    new_op.apply_to_normal_form(input);
                }

                Op::Overlay { operations } => {
                    let mut result = vec![];
                    for op in operations {
                        let mut input_clone = input.clone();
                        op.apply_to_normal_form(&mut input_clone);
                        result.append(&mut input_clone);
                    }

                    *input = result
                }
            }

            match_length(input);
        }
    }

    fn match_length(input: &mut NormalForm) {
        let max_len = get_max_length_ratio(&input);
        for voice in input {
            let mut voice_len = Ratio::new(0, 1);
            for op in voice.clone() {
                voice_len += op.get_length_ratio()
            }
            if voice_len < max_len {
                voice.push(Silence {
                    m: max_len - voice_len,
                });
            }
        }
    }

    fn get_max_length_ratio(input: &NormalForm) -> Rational {
        let mut max_len = Ratio::new(0, 1);
        for voice in input {
            let mut voice_len = Ratio::new(0, 1);
            for op in voice {
                voice_len += op.get_length_ratio()
            }

            if voice_len > max_len {
                max_len = voice_len
            }
        }

        max_len
    }

    fn join_sequence(mut l: NormalForm, mut r: NormalForm) -> NormalForm {
        if l.len() == 0 {
            return r;
        }

        let diff = l.len() as isize - r.len() as isize;
        let l_max_len = get_max_length_ratio(&l);
        let r_max_len = get_max_length_ratio(&r);
        match diff.partial_cmp(&0).unwrap() {
            Equal => {}
            Greater => {
                for _ in 0..(diff.abs()) {
                    r.push(vec![Op::Silence { m: r_max_len }])
                }
            }
            Less => {
                for _ in 0..diff.abs() {
                    l.push(vec![Op::Silence { m: l_max_len }])
                }
            }
        }

        let mut result = vec![];
        for (left, right) in l.iter_mut().zip(r.iter_mut()) {
            left.append(right);

            result.push(left.clone())
        }

        result
    }
}
