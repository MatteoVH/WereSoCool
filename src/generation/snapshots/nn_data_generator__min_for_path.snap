---
created: "2019-05-13T22:07:31.887680Z"
creator: insta@0.8.1
source: src/generation/nn_data_generator.rs
expression: min_state
---
CSVOp {
    fm: 0.0,
    fa: -10.2,
    pm: -1.0,
    pa: -1.7857142857142858,
    g: 0.0,
    l: 0.0,
    v: 0
}
